File: 31-next-permutation/next-permutation.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function (a) {
    let ind = -1;
    let n = a.length;

    for (i = n - 2; i >= 0; i--) {
        if (a[i] < a[i + 1]) {
            ind = i;
            break;
        }
    }

    if(ind == -1 ){
        return a.reverse(); 
    }

    for (i = n - 1 ; i >= ind; i--) {
        if(a[i]>a[ind]){
            let temp = a[i]
            a[i] = a[ind]
            a[ind] = temp;

            break;
        }
    }

    a.splice(ind + 1, n - ind - 1, ...a.slice(ind + 1).reverse());

    return a
};

File: 31-next-permutation/next-permutation.java
class Solution {
    void swap(int a, int b, int[] nums) {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }

    void rev(int start, int end, int[] nums) {
        while (start <= end) {
            swap(start, end, nums);
            start++;
            end--;
        }
    }

    public void nextPermutation(int[] nums) {
        int ind = -1;
        int n = nums.length;

        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                ind = i;
                break;
            }
        }

        if (ind == -1) {
            rev(0, n - 1, nums);
            return;
        }

        for (int i = n - 1; i >= 0; i--) {
            if (nums[i] > nums[ind]) {
                swap(i, ind, nums);
                break;
            }
        }

        rev(ind + 1, n - 1, nums);
    }
}